### PR TITLE
Avoid scanning twice tsconfig.json project references

### LINKED ISSUE #3556: Avoid scanning twice `tsconfig.json` project references
tsconfig returned references should include `tsconfig.json` paths to avoid duplicating the paths (and thus scanning twice) on Java workList

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**
Fixes #3556

### CODE REVIEW

File: eslint-bridge/src/services/program/program.ts [hunk 1]
```diff
@@ -149,22 +148,36 @@ export async function createProgram(inputTsConfig: string): Promise<{
...
   const program = ts.createProgram(programOptions);
-  const maybeProjectReferences = program.getProjectReferences();
-  const projectReferences = maybeProjectReferences ? maybeProjectReferences.map(p => p.path) : [];
+  const projectReferences = program.getProjectReferences() || [];
+  const sanitizedProjectReferences: string[] = [];
+
+  for (const reference of projectReferences) {
+    const sanitizedReference = await addTsConfigIfMissing(reference.path);
+    if (!sanitizedReference) {
+      console.log(`WARN Skipping missing referenced tsconfig: ${reference.path}`);
```
Comment (vilchik-elena) [inline eslint-bridge/src/services/program/program.ts:164]: I suggest in the logs we either use `tsconfig.json` or `TSConfig` and not just `tsconfig`

File: eslint-bridge/tests/services/program/program.test.ts [hunk 2]
```diff
@@ -44,7 +44,19 @@ describe('program', () => {
         toUnixPath(path.join(reference, 'file.ts')),
       ]),
     );
-    expect(projectReferences).toEqual([toUnixPath(reference)]);
+    expect(projectReferences).toEqual([await addTsConfigIfMissing(reference)]);
```
Comment (vilchik-elena) [inline eslint-bridge/tests/services/program/program.test.ts:47]: I prefer avoid using `addTsConfigIfMissing` and use `path.join` instead to be explicit about the expectation

File: eslint-bridge/tests/services/program/program.test.ts [hunk 3]
```diff
@@ -44,7 +44,19 @@ describe('program', () => {
...
+  });
+
+  it('should skip missing reference a program', async () => {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/services/program/program.test.ts:50]: not sure about the grammar of this phrase :)

File: eslint-bridge/src/services/program/program.ts [hunk 4]
```diff
@@ -149,22 +148,36 @@ export async function createProgram(inputTsConfig: string): Promise<{
   projectReferences: string[];
   missingTsConfig: boolean;
```
Comment (vilchik-elena) [inline eslint-bridge/src/services/program/program.ts:149]: as now indeed it's ambiguous why we don't set this property to true for a missing reference, I suggest to add jsdoc to this function explaining that it's true when extended tsconfig is missing and we fallback to a default config.

File: eslint-bridge/tests/services/program/program.test.ts [hunk 5]
```diff
@@ -151,9 +163,13 @@ describe('program', () => {
     expect(program.getRootFileNames()).toEqual(
       files.map(toUnixPath).filter(file => file.startsWith(toUnixPath(fixtures))),
     );
-    expect(program.getProjectReferences().map(reference => reference.path)).toEqual(
-      projectReferences,
-    );
+    expect(
```
Comment (vilchik-elena) [inline eslint-bridge/tests/services/program/program.test.ts:166]: I wonder why we even care about this assertion, I think we never rely on `program.getProjectReferences()`.
Even if we keep it, is it possible to simplify asyncs/awaits?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/helpers/files.ts [hunk 6]
```diff
@@ -74,3 +75,20 @@ export function stripBOM(str: string) {
...
+ */
+export async function addTsConfigIfMissing(tsConfig: string) {
+  try {
+    if ((await fs.lstat(tsConfig)).isDirectory()) {
+      tsConfig = path.join(tsConfig, 'tsconfig.json');
+    }
+
+    return tsConfig;
+  } catch {
+    return null;
```
[the remark this answers is not in the document]
Comment (victor-diez-sonarsource) [reply eslint-bridge/src/helpers/files.ts:92]: yes, we can try catch here or on the function call. if not here, we will need to handle promise rejections, or try...catchs. What do you prefer?

File: eslint-bridge/src/services/program/program.ts [hunk 7]
```diff
@@ -149,22 +148,36 @@ export async function createProgram(inputTsConfig: string): Promise<{
...
   const programOptions = createProgramOptions(tsConfig);
 
   const program = ts.createProgram(programOptions);
-  const maybeProjectReferences = program.getProjectReferences();
-  const projectReferences = maybeProjectReferences ? maybeProjectReferences.map(p => p.path) : [];
+  const projectReferences = program.getProjectReferences() || [];
+  const sanitizedProjectReferences: string[] = [];
+
+  for (const reference of projectReferences) {
+    const sanitizedReference = await addTsConfigIfMissing(reference.path);
```
[the remark this answers is not in the document]
Comment (victor-diez-sonarsource) [reply eslint-bridge/src/services/program/program.ts:162]: agreed, we had this conversation, remember? :P

File: its/plugin/projects/referenced-tsconfigs/dir/tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "files": ["file.ts"],
+  "references": [
+    {
+      "path": "..\\tsconfig.json"
```
[the remark this answers is not in the document]
Comment (victor-diez-sonarsource) [reply its/plugin/projects/referenced-tsconfigs/dir/tsconfig.json:5]: checking windows paths in tsconfig.json. seems they are supported, so I wanted to test proper behaviour

File: eslint-bridge/tests/services/program/program.test.ts (hunk 5 above)
Comment (victor-diez-sonarsource) [reply eslint-bridge/tests/services/program/program.test.ts:166]: ok, remove it. references are already tested in other tests

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**